lib/datadog/tracing/contrib/rack/middlewares.rb

```diff
@@ -93,13 +96,27 @@ def call(env)
             request_span.resource = nil
             request_trace = Tracing.active_trace
             env[Ext::RACK_ENV_REQUEST_SPAN] = request_span
+            Thread.current[:datadog_http_routing] = []
 
             # Copy the original env, before the rest of the stack executes.
             # Values may change; we want values before that happens.
             original_env = env.dup
 
             # call the rest of the stack
             status, headers, response = @app.call(env)
+
+            if status != 404 && (routed = Thread.current[:datadog_http_routing].last)
+              last_script_name = routed[1]
+              last_route = routed[2]
+
+              if last_script_name == '' && env['SCRIPT_NAME'] != ''
```

Comment:
Can we add a comment explaining about such exceptional case?

Reply:
I can't seem to wrap my head around it either.

In the conditioned code:

- `last_route` would always be `''`
- `last_script_name = env['PATH_INFO']` means there's presumably interesting info in `PATH_INFO` but what of the `env['SCRIPT_NAME']` value, which is `!= ''`?

I would also be wary of what the `env['PATH_INFO']` and `env['SCRIPT_NAME']` values would be assumed to be in case of nesting: if they've been mutated at each stage of nesting they may not be the original values at the beginning of the request. This may or may not be what's intended to be relied upon here, but that's very unclear.

Also I can't recall what happens when e.g NGiNX is in front and proxying under a path, e.g NGiNX `example.com/foo/(.*)` is proxied to `app:3000/$1` (`SCRIPT_NAME` may be already set to `/foo` CGI-style - that's the part I can't recall - even at the very top of the Rack stack)

Reply:
> if they've been mutated at each stage of nesting they may not be the original values

At least Rack `map` ensures they're being restored to the previous value. Dunno about other cases.

https://github.com/rack/rack/blob/3897649e8740e560a5fa142f972121a119b26b5c/lib/rack/urlmap.rb#L81-L84

Reply:
> last_route would always be ''

I changed the struct of the item in `Thread.current` so that is in the form {integration, full_path, integration_route} where full_path -> env['SCRIPT_INFO'] and integration_route -> the end route value according to that integration. So last_route will not be '' ever except for the exception case detailed below

Reply:
Update with comment but quick summary is that for request of nested rack requests ONLY, we essentially get just the root of the first rack request and all other nested rack requests are ignored. Leading to things like the route of `/rack/rack/hello/world` to be `/rack` which is incorrect. This fixes it by ensuring that there exists a route and appending it only when certain conditions are met in a nested rack request ONLY